ESQL: introduce support for mapping-unavailable fields

bpintea opened the pull request:
This introduces support for mapping-unavailable fields (present and not mapped or just missing). The behaviour is controlled through a new SET setting `unmapped_fields`, which can take the values "FAIL", "NULLIFY", "LOAD".

An optional field behaves just like a "normal", mapped field, with regards to how it flows through the commands chain: it can be simply used in the commands, as if present in the source, but can no longer be referenced once dropped - explicitly, with `DROP`, or not selected by a `KEEP`, or `RENAME` that doesn't reference it -,  or past a `STATS` reduction. 

However, unlike a mapped field, if it's not reference at all, it won't show up in the output of a simple `FROM index`.

Currently, the schema difference between nullified fields and the loaded ones is in the type: nullified ones are of data type  `NULL`, while the loaded ones are `KEYWORD`.
The implementation difference w.r.t. logical plan building is that the nullified fields are created as `null` value aliasing on top of the data source, while the loaded one are pushed as extractors into the source (this leverages the [INSIST work](https://github.com/elastic/elasticsearch/pull/119886)).

The partially mapped fields are also covered: when the setting is "load", these fields will be extracted from those indices that have the field, but isn't mapped. In case there's a conflict between the loaded `KEYWORD` field and the mapped type in the fields that have this field mapped, an explicit conversion is needed, just like with union types.

Related: #138888

bpintea committed:
add support for unmapped_fields=nullify

Commit:
[CI] Auto commit changes from spotless

bpintea committed:
fix EXPLAIN

bpintea committed:
Merge remote-tracking branch 'upstream/main' into feat/set_unmapped_nullify

bpintea committed:
fix docs

bpintea committed:
Merge remote-tracking branch 'upstream/main' into feat/set_unmapped_nullify

Commit:
[CI] Auto commit changes from spotless

bpintea committed:
add support for UnionAll, Fork

bpintea committed:
Merge remote-tracking branch 'upstream/main' into feat/set_unmapped_nullify

bpintea committed:
flip setting snapshot-only flag to true

Commit:
[CI] Auto commit changes from spotless

bpintea committed:
more tests

bpintea committed:
add spec tests

bpintea committed:
Merge remote-tracking branch 'upstream/main' into feat/set_unmapped_nullify

bpintea committed:
fix one test

bpintea committed:
Merge remote-tracking branch 'upstream/main' into feat/set_unmapped_nullify

bpintea committed:
Update docs/changelog/139417.yaml

bpintea committed:
test stabilization

bpintea committed:
Merge remote-tracking branch 'upstream/main' into feat/set_unmapped_nullify

bpintea committed:
update test failing on unrelated error

bpintea committed:
subquery tests deferred

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:
```diff
@@ -2142,6 +2212,130 @@ private static Expression typeSpecificConvert(ConvertFunction convert, Source so
         }
     }
 
+    private static class ResolveUnmapped extends ParameterizedAnalyzerRule<LogicalPlan, AnalyzerContext> {
```
thought: as the analyzer grows, we should probably start putting analyzer rules into their own files. It's fine for this PR, of course.

bpintea replied:
Yes, I had the same thought even before this PR; we do the same with the optimizer rules, don't think we need this huge class. I've applied the suggestion. I will follow up with breaking the class down (at some point).

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:
```diff
@@ -2142,6 +2212,130 @@ private static Expression typeSpecificConvert(ConvertFunction convert, Source so
         }
     }
 
+    private static class ResolveUnmapped extends ParameterizedAnalyzerRule<LogicalPlan, AnalyzerContext> {
```
That class would probably profit from a javadoc.

bpintea replied:
Added.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:
```diff
@@ -2142,6 +2212,130 @@ private static Expression typeSpecificConvert(ConvertFunction convert, Source so
+            };
+        }
+
+        private LogicalPlan nullify(LogicalPlan plan) {
```
It's really unclear what to expect from this method without a short high-level explanation of the approach/algorithm in a javadoc.

bpintea replied:
I've added further javadoc.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:
```diff
@@ -2142,6 +2212,130 @@ private static Expression typeSpecificConvert(ConvertFunction convert, Source so
+            List<UnresolvedAttribute> nullifiedUnresolved = new ArrayList<>();
+            List<UnresolvedAttribute> unresolved = new ArrayList<>();
+            Holder<Boolean> unresolvable = new Holder<>(false);
+            var resolved = plan.transformDown(p -> unresolvable.get() ? p : switch (p) {
```
Why do we go down the whole plan? Because we're inside a ParameterizedAnalyzerRule, we will already have walked all plan children before reaching here.

Using ResolveRefs as a comparison, I'd expect something like:
- check if the plan has unresolved children; if so, skip (children has attributes that couldn't be nullified for whatever reason)
- check for new unresolved attributes in the current plan node, only
- if any are found, do whatever we need to do to nullify them.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:
```diff
@@ -2142,6 +2212,130 @@ private static Expression typeSpecificConvert(ConvertFunction convert, Source so
+            }
+            nullifiedUnresolved.addAll(unresolved);
+            unresolved.clear(); // cleaning since the plan might be n-ary, with multiple sources
+            return new Eval(p.source(), p, List.copyOf(aliasesMap.values()));
```
The opposite may also happen, no? That is, we inject an `EVAL missing = null`, but a downstream node shadows or drops it. That's more okay, though, because it follows the semantic of "we act as if the index had the field, and if the query goes and drops it, oh well."

bpintea replied:
This has been refactored.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/core/expression/UnresolvedAttribute.java:
```diff
@@ -97,14 +97,18 @@ protected Attribute clone(
-    }
+        // assertions, "should never happen"
+        if (dataType != null) {
+            throw new UnresolvedException("clone() with data type [" + dataType + "]", this);
```
nit: Let's throw ISE if that shouldn't happen?

Maybe we could also skip this if assertions aren't on?

alex-spies reviewed (commented):
Heya!

I _only_ reviewed the approach (focusing mostly on ResolveUnmapped) and the changes to the attribute hierarchy.

The approach seems right to me. I think the semantics are correct. We're going to have a lot of fun getting renames and `KEEP missing*` to work - esp. with potential shadowing! - but I think it's right that the `NULLIFY` settings makes us treat a missing field as if it was, in fact, an indexed field (just one that's constantly `null`). The alternative would be to inject nulls in any place where a field is missing; that'd be easier, but would probably not mesh so well with the semantics we'll need for the `LOAD` setting - which will force us to deal with such nastiness anyway.

My comments focus mostly on clarity; this is not a simple change, and I found it difficult to follow, so I think increasing clarity is important for our future selves.

GalLalouche commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/UnmappedResolution.java:
```diff
@@ -0,0 +1,14 @@
+
+package org.elasticsearch.xpack.esql.analysis;
+
+public enum UnmappedResolution {
```
Needs a javadoc (Or a link to a place where these meanings are further explained).

bpintea replied:
Added.

GalLalouche commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/parser/SetParserTests.java:
```diff
@@ -162,4 +167,31 @@ private Object settingValue(EsqlStatement query, int position) {
+        var mode = randomValueOtherThanMany(
+            v -> Arrays.stream(UnmappedResolution.values())
+                .map(x -> x.name().toLowerCase(Locale.ROOT))
+                .toList()
```
Not that it matter much for tests on an `enum` of length 3, but in general, this is slightly cleaner and faster :)
```java
v -> Arrays.stream(UnmappedResolution.values()).anyMatch(x -> x.name().equalsIgnoreCase(v))
```

GalLalouche commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/analysis/AnalyzerTestUtils.java:
```diff
@@ -168,11 +210,26 @@ public static LogicalPlan analyze(String query, TransportVersion transportVersio
+    private static final Map<String, EsField> MAPPING_BASIC_RESOLUTION = EsqlTestUtils.loadMapping("mapping-basic.json");
+
+    private static Map<IndexPattern, IndexResolution> indexResolutions(@Nullable String indexName) {
+        Map<IndexPattern, IndexResolution> indexResolutions;
```
I personally find the pattern of
```java
Type value;
if (cond) {
    value = something;
} else {
    value = somethingElse;
}
return value;
```
To be a lot less readable than a ternary operator or just an early return, but that might be my functional upbringing 🙃. Consider:
```java
        if (indexName == null) {
            return Map.of();
        }
        var indexResolution = IndexResolution.valid(
            new EsIndex(indexName, MAPPING_BASIC_RESOLUTION, Map.of(indexName, IndexMode.STANDARD), Map.of(), Map.of(), Set.of())
        );
        return Map.of(new IndexPattern(Source.EMPTY, indexName), indexResolution);
```
Or even:
```java
        return indexName == null ? Map.of() : Map.of(
            new IndexPattern(Source.EMPTY, indexName),
            IndexResolution.valid(
                new EsIndex(indexName, MAPPING_BASIC_RESOLUTION, Map.of(indexName, IndexMode.STANDARD), Map.of(), Map.of(), Set.of())
            )
        );
```

bpintea replied:
Went for the former, the latter is too "busy". I do prefer the ternary generally too.

GalLalouche commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/core/expression/Attribute.java:
```diff
@@ -181,6 +183,10 @@ protected abstract Attribute clone(
         boolean synthetic
     );
 
+    private DataType safeDataType() {
```
1. Please mark with `@Nullable`.
2. Consider renaming to something like `resolvedDataType` or `dataTypeIfResolved`.

GalLalouche commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:
```diff
@@ -1261,7 +1272,62 @@ private LogicalPlan resolveKeep(Project p, List<Attribute> childOutput) {
+        }
+
+        /**
+         * Other rules (like {@link ResolveUnmapped}) will further resolve attributes that this rule will not, in a first pass.
```
Javadoc is pretty hard to parse :) Perhaps rephrase?

bpintea replied:
Rephrased.

bpintea committed:
Apply review comments

bpintea committed:
Merge remote-tracking branch 'upstream/main' into feat/set_unmapped_nullify

bpintea commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:
```diff
@@ -2142,6 +2212,130 @@ private static Expression typeSpecificConvert(ConvertFunction convert, Source so
+                case EsRelation relation -> evalUnresolved(relation, unresolved, nullifiedUnresolved);
+                // each subquery aliases its own unresolved attributes "internally" (before UnionAll)
+                case Fork fork -> {
+                    if (fork.output().isEmpty()) {
```
Reworked this away.

bpintea commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/analysis/AnalyzerTestUtils.java:
```diff
@@ -135,7 +166,18 @@ public static Analyzer analyzer(Map<IndexPattern, IndexResolution> indexResoluti
+        var statement = EsqlParser.INSTANCE.createStatement(query);
+        var relations = statement.plan().collectFirstChildren(UnresolvedRelation.class::isInstance);
+        var indexName = relations.isEmpty() ? null : ((UnresolvedRelation) relations.getFirst()).indexPattern().indexPattern();
+        var indexResolutions = indexResolutions(indexName);
```
Same here :)

bpintea commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:
```diff
@@ -2142,6 +2212,130 @@ private static Expression typeSpecificConvert(ConvertFunction convert, Source so
+            };
+        }
+
+        private LogicalPlan nullify(LogicalPlan plan) {
```
Fixed, thx.

bpintea commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:
```diff
@@ -2142,6 +2212,130 @@ private static Expression typeSpecificConvert(ConvertFunction convert, Source so
+                case Project project -> { // TODO: redo handling of `KEEP *`, `KEEP foo* | EVAL foo_does_not_exist + 1`, `RENAME ` etc.
+                    if (hasUnresolvedNamedExpression(project)) {
+                        unresolvable.set(true); // `UnresolvedNamedExpression`s need to be turned into attributes first
+                        yield p; // give up
```
This got refactored away.

bpintea commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/core/expression/Attribute.java:
FTR, these changes got reverted (though they're themselves applicable, but are no longer required here and they're kid of relevant).

bpintea committed:
Add support for loading fields

bpintea committed:
Merge remote-tracking branch 'upstream/main' into feat/set_unmapped_nullify

bpintea committed:
add support for partially mapped fields

GalLalouche commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/UnmappedResolution.java:
```diff
@@ -0,0 +1,31 @@
+    NULLIFY,
+
+    /**
+     * Just like {@code NULLIFY}, but instead of null-aliasing, insert extractors in the data source.
```
1. (In general) Use {@link NULLIFY} to link to the `enum` variant directly, but see below.
2. This is very much *unlike* `NULLIFY`, given that it actually (attempts) to load the data... Perhaps rephrase to:
```java
In case the query references a field that's not present in the index mapping, attempts to load it from {@code _source}."
```

GalLalouche commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/rules/ResolveUnmapped.java:
```diff
@@ -0,0 +1,295 @@
+ * The rule handles fields that don't show up in the index mapping, but are used within the query. These fields can either be missing
+ * entirely, or be present in the document, but not in the mapping (which can happen with non-dynamic mappings).
+ * <p>
+ * In the case of the former ones, the rule introducees {@code EVAL missing = NULL} commands (null-aliasing / null-Eval'ing).
```
1. s/introducees/introduces.
2. This comment is misleading, since it reads like the rule chooses nullifying/loading based on the data being present or not, and not based on the `enum` passed to it, which is configured by the user.

bpintea replied:
1. Fixed, thx.
2. Added a clarification.

GalLalouche commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/rules/ResolveUnmapped.java:
```diff
@@ -0,0 +1,295 @@
+        // insert an Eval on top of those LeafPlan that are children of n-ary plans (could happen with UnionAll)
+        transformed = transformed.transformUp(
+            n -> n instanceof UnaryPlan == false && n instanceof LeafPlan == false,
+            nAry -> evalUnresolved(nAry, nullAliases)
```
While the overload here does "the right thing", I find it pretty confusing that it only works due the child being more specific. Consider just using two different names here, e.g., `evalUnresolveUnary`.

bpintea replied:
Updated.

GalLalouche commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/rules/ResolveUnmapped.java:
```diff
@@ -0,0 +1,295 @@
+    }
+
+    private static void assertSourceType(LogicalPlan source) {
+        switch (source) {
```
Personally, I think this part would be simpler with a plain old `if (cond) { throw }`.

bpintea replied:
The `cond` would be hard to read: `source instance of EsRelation == false &&` repeated two more times.

GalLalouche commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/rules/ResolveUnmapped.java:
```diff
@@ -0,0 +1,295 @@
+    private static List<Alias> nullAliases(List<UnresolvedAttribute> unresolved) {
+        Map<String, Alias> aliasesMap = new LinkedHashMap<>(unresolved.size());
+        for (var u : unresolved) {
+            if (aliasesMap.containsKey(u.name()) == false) {
```
Consider using `putIfAbsent`, or if you want to avoid the allocation: `computeIfAbsent`.

GalLalouche reviewed (approved):
I've left a few small style comments/suggestions, but otherwise LGTM!

bpintea commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/rules/ResolveUnmapped.java:
```diff
@@ -0,0 +1,295 @@
+            Holder<Boolean> patched = new Holder<>(false);
+            child = child.transformDown(
+                // TODO add a suitable forEachDownMayReturnEarly equivalent
+                n -> patched.get() == false && n instanceof Project, // process top Project only (Fork-injected)
```
There's a tiny subtlety here: I want to skip the type checks for the remainder of the tree once the boolean flag flips.

bpintea commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/rules/ResolveUnmapped.java:
```diff
@@ -0,0 +1,295 @@
+        // insert an Eval on top of every LeafPlan, if there's a UnaryPlan atop it
+        var transformed = plan.transformUp(
+            n -> n instanceof UnaryPlan unary && unary.child() instanceof LeafPlan,
+            p -> evalUnresolved((UnaryPlan) p, nullAliases)
```
Kept this, however : I find it "clarifying" to see in one place the distinction between the two cases, rather than delegate the checks into the methods.

bpintea committed:
Address review comments

bpintea committed:
Merge remote-tracking branch 'upstream/main' into feat/set_unmapped_nullify

bpintea commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/rules/ResolveUnmapped.java:
```diff
@@ -0,0 +1,295 @@
+     */
+    private static LogicalPlan load(LogicalPlan plan, List<UnresolvedAttribute> unresolved) {
+        // TODO: this will need to be revisited for non-lookup joining or scenarios where we won't extraction from specific sources
+        var transformed = plan.transformUp(n -> n instanceof EsRelation esr && esr.indexMode() != IndexMode.LOOKUP, n -> {
```
Ah, yes, updated.
When I was writing it I thought I might want to handle other types of sources (like maybe `LocalRelation`), but that not needed (since we don't deal with that in the analyzer).

bpintea commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/UnmappedResolution.java:
```diff
@@ -0,0 +1,31 @@
+import org.elasticsearch.xpack.esql.core.type.DataType;
+
+/**
+ * This is a unmapped-fields strategy discriminator.
```
Nit 1: addressed. 
Nit 2: kept, as this is what an editor will display when showing info about the class/enum.

nik9000 commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/unmapped-load.csv-spec:
```diff
@@ -0,0 +1,636 @@
+
+SET unmapped_fields="load"\;
+FROM partial_mapping_excluded_source_sample_data
+| SORT @timestamp DESC
```
If you want `message` you can have it with `KEEP message`, right? Or you could do something to do and it'd "appear", right?

astefan reviewed (commented):
There is some very good work in this PR, especially the attention to the different use cases and commands that can use "unmapped fields". Good job with this and the tests; I learned a bit about this feature by looking and trying out different queries from tests.

I am seeing difficulties in adding such a broad feature with different tricky/special commands to account for (`fork`, sub-queries, _insist and union types) and I am seeing some inconsistent behavior and outcome (see my comments). I don't think I have enough context to understand the full list of functionality for this feature, but I am seeing some things missing and I am curious on the plan for those missing bits/limitations (maybe?).

I will go over the PR one more time, at least. But, so far it is looking good, ignoring the missing bits/issues I mentioned above.

bpintea committed:
Introduce a ResolvingProject

This will allow re-evaluating the output past a RENAME/DROP/KEEP, once
an unmapping field is injected.

bpintea committed:
Merge remote-tracking branch 'upstream/main' into feat/set_unmapped_nullify

Commit:
[CI] Auto commit changes from spotless

bpintea committed:
fix node tests

bpintea committed:
Merge remote-tracking branch 'upstream/main' into feat/set_unmapped_nullify

bpintea committed:
Revert ResolvingProject

bpintea committed:
Merge remote-tracking branch 'upstream/main' into feat/set_unmapped_nullify

bpintea commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/core/expression/UnresolvedPattern.java:
> Maybe that's wrong, and we should be able to stick unresolved messages also on UnresolvedNamePattern; or maybe we need something like an UnresolvableNamePattern that implements Unresolvable?

Currently, `UnresolvedNamePattern` are just turned into a `UnresolvedAttribute` if they match nothing at `ResolveRefs` time.
We need something that's a subclass of `Attribute` here. The naming could be improved, but we simply need something akin `UnresolvedAttribute`, that doesn't look like an unmapped field (which is also an `UnresolvedAttribute`).

bpintea commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/unmapped-nullify.csv-spec:
```diff
@@ -0,0 +1,373 @@
+ROW x = 1
+| EVAL language_code = does_not_exist::INTEGER
+| LOOKUP JOIN languages_lookup ON language_code
+;
```
That is not, there's just one `SET` allowed per statment/query.

bpintea committed:
Reintroduce ResolvingProject

bpintea committed:
Merge remote-tracking branch 'upstream/main' into feat/set_unmapped_nullify

astefan reviewed (approved):
LGTM. Left some comments exploratory in nature and, if they are valid, a potential set of improvements. But can be addressed later with further work on this feature. Great work in this PR!

--- 

A comment for later discussions (if any), not a show stopper, more like a heads up for some potential confusion amongst users and weird hard-to-track-down bugs later in the development cycles.

The `unmapped_message` is materialized whenever `KEEP` is used. If I do `FROM partial_mapping_sample_data | SORT unmapped_message` I get an error about the field not being found. That's fair and **this shows that there should be a mechanism for some "special" fields to be made available in a query command**. I am also aware that this should improve with follow-up PRs.

If these "special" fields (unmapped, not visible etc) need a separate mechanism to be made available, why are we treating them equal to other fields, the "normal" ones? IMHO, to have a clean story around these unmapped fields, we need to "signal" their presence in a different way that shouldn't lead users to think that they are treated equally. With `_insist` command (that is snapshot only currently) **this intention is very clear**: `_insist` is a separate mechanism by which a user tells ESQL that whatever field name `_insist` refers to, that field is "special" in some way or another.

At some point, I _think_ this transparent way of loading unmapped fields with `keep` will come back and bite us because these fields are not like the regular fields. We shouldn't treat them as regular fields, to eliminate any sort of confusion. This could also cripple our work on optimizations where we have several mechanisms to add `eval`s to account for possible shadowed fields, where we move (potentially eliminate) commands around. This will make it easier to introduce bugs hard to track down and fix later in the development cycle.

Looking at how these fields are used and loaded and because they depend on being `KEEP`ed in some way or another, I am wondering why the explicit reference to a potentially unmapped field is not part of `FROM`.

Just a humble 2c.

alex-spies commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/unmapped-load.csv-spec:
```diff
@@ -0,0 +1,651 @@
+2024-10-23T13:55:01.543Z | partial_mapping_sample_data                 | Connected to 10.1.0.1!
+;
+
+partialMappingUnionTypes
```
pre-existing but misleading: there are no union types involved here. The only thing that happens is that we pass `message` to `to_string` first thing, but there's no mapping conflict to be solved.

GalLalouche replied:
Renamed and replaced `::STRING` with the less confusing `::KEYWORD` in #143522.

alex-spies reviewed (commented):
Gave it another go. Need to do at least another round, this has become _huge_.

Ideas for follow-up work, summarized from my comments below:
- We may want to issue a warning if a user sets `unmapped_fields:"load"` but the source for one of the indices is disabled.
- [We need to figure](https://github.com/elastic/elasticsearch/pull/139417#discussion_r2658270134) out how to avoid having both `UnresolvedNamePattern` and `UnresolvedPattern`. I think there's a way, and in the current state, the attribute/named expression hierarchy is pretty confusing with these two co-existing.
- Our generative tests should probably randomly prepend one of the 3 settings for `unmapped_fields`.
- We want more tests for `unmapped_fields=\"load\"`. E.g.
  - Tests for the behavior of union types when an index has an unmapped/missing field that is multi-typed based on the mapping of 2 other indices.
  - More tests that do something with the unmapped/missing fields (e.g. join or enrich on them!)
  - More tests that do something _before_ using the unmapped fields, esp. join on an unrelated field.
  - Subqueries using an unmapped/missing field

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/EsqlCapabilities.java:
```diff
@@ -148,11 +148,17 @@ public enum Cap {
+        /**
+         * Support for optional fields (might or might not be present in the mappings).
+         */
+        OPTIONAL_FIELDS,
```
This capability should be SNAPSHOT-only. We plan to un-snapshot in stages, starting with the `nullify` setting; that's better left for a follow-up PR.

Let's also slap `test-release` on this, so we make sure not to break the build.

bpintea commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/rules/ResolveUnmapped.java:
```diff
@@ -0,0 +1,317 @@
+    private static LogicalPlan nullify(LogicalPlan plan, List<UnresolvedAttribute> unresolved) {
+        var nullAliases = nullAliases(unresolved);
+
+        // insert an Eval on top of every LeafPlan, if there's a UnaryPlan atop it
```
No, thanks, this is now [being merged](https://github.com/elastic/elasticsearch/pull/141340/changes#diff-b135751577c8077e6aebc28a662beca9f604826869d1d368c60fc8f1bb0a2e4bR99-R109).

bpintea commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/rules/ResolveUnmapped.java:
```diff
@@ -0,0 +1,317 @@
+                // (A `new UnresolvedAttribute(ua.source(), ua.name())` would save an allocation, but is problematic with subtypes.)
+                ua = (ua.withId(new NameId())).withUnresolvedMessage(null);
+            }
+            return ua;
```
They could probably have been, though it might not have been a clear code. But refreshing the plan has been factored away.

bpintea commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/rules/ResolveUnmapped.java:
```diff
@@ -0,0 +1,317 @@
+        return insisted;
+    }
+
+    // TODO: would an alternative to this be to drop the current Fork and have ResolveRefs#resolveFork re-resolve it. We might need
```
Yes, I think we should probably refactor Fork for more robustness (a similar [comment](https://github.com/elastic/elasticsearch/pull/140463#discussion_r2745862750)).

GalLalouche commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/unmapped-load.csv-spec:
```diff
@@ -0,0 +1,651 @@
+;
+
+#########################
+# Multi-parameter tests #
```
Fixed in #143522.
